File: bins/astra-sim/src/protocol.rs
//! CT002 UDP protocol — standalone copy of the frame builder/parser the
//! simulator uses to talk to an astrameter CT002 emulator. Mirrors
//! `src/astrameter/simulator/protocol.py` byte-for-byte (separator,
//! checksum, length encoding, firmware quirk tolerance).

pub const SOH: u8 = 0x01;
pub const STX: u8 = 0x02;
pub const ETX: u8 = 0x03;
pub const SEPARATOR: char = '|';

#[allow(dead_code)]
pub const RESPONSE_LABELS: &[&str] = &[
    "meter_dev_type",
    "meter_mac_code",
    "hhm_dev_type",
    "hhm_mac_code",
    "A_phase_power",
    "B_phase_power",
    "C_phase_power",
    "total_power",
    "A_chrg_nb",
    "B_chrg_nb",
    "C_chrg_nb",
    "ABC_chrg_nb",
    "wifi_rssi",
    "info_idx",
    "x_chrg_power",
    "A_chrg_power",
    "B_chrg_power",
    "C_chrg_power",
    "ABC_chrg_power",
    "x_dchrg_power",
    "A_dchrg_power",
    "B_dchrg_power",
    "C_dchrg_power",
    "ABC_dchrg_power",
];

/// Phase-letter → request-field index. Mirrors Python's `PHASE_FIELD_INDEX`.
pub fn phase_field_index(p: char) -> Option<usize> {
    match p {
        'A' => Some(4),
        'B' => Some(5),
        'C' => Some(6),
        _ => None,
    }
}

pub fn calculate_checksum(data: &[u8]) -> u8 {
    data.iter().fold(0u8, |acc, b| acc ^ *b)
}

pub fn compute_length(payload_without_length: &[u8]) -> usize {
    let base_size = 1 + 1 + payload_without_length.len() + 1 + 2;
    for length_digits in 1..5 {
        let total = base_size + length_digits;
        if total.to_string().len() == length_digits {
            return total;
        }
    }
    panic!("payload length too large");
}

pub fn build_payload(fields: &[String]) -> Vec<u8> {
    let mut message_str = String::new();
    message_str.push(SEPARATOR);
    for (i, f) in fields.iter().enumerate() {
        if i > 0 {
            message_str.push(SEPARATOR);
        }
        message_str.push_str(f);
    }
    let message_bytes = message_str.into_bytes();
    let total_length = compute_length(&message_bytes);

    let mut payload = Vec::with_capacity(total_length);
    payload.push(SOH);
    payload.push(STX);
    payload.extend(total_length.to_string().as_bytes());
    payload.extend_from_slice(&message_bytes);
    payload.push(ETX);
    let cs = calculate_checksum(&payload);
    payload.extend(format!("{cs:02x}").as_bytes());
    payload
}

/// Parse a CT002 response message. Returns `(fields, None)` on success,
/// `(None, Some(reason))` on failure.
pub fn parse_message(data: &[u8]) -> (Option<Vec<String>>, Option<String>) {
    if data.len() < 10 {
        return (None, Some("Too short".into()));
    }
    if data[0] != SOH || data[1] != STX {
        return (None, Some("Missing SOH/STX".into()));
    }
    let sep_index = match data[2..].iter().position(|b| *b == b'|') {
        Some(i) => i + 2,
        None => return (None, Some("No separator after length".into())),
    };
    let length: usize = match std::str::from_utf8(&data[2..sep_index])
        .ok()
        .and_then(|s| s.parse().ok())
    {
        Some(n) => n,
        None => return (None, Some("Invalid length field".into())),
    };
    if data.len() != length {
        return (
            None,
            Some(format!(
                "Length mismatch (expected {length}, got {})",
                data.len()
            )),
        );
    }
    if data[length - 3] != ETX {
        return (None, Some("Missing ETX".into()));
    }
    let xor = data[..length - 2].iter().fold(0u8, |acc, b| acc ^ *b);
    let expected = format!("{xor:02x}");
    let actual = &data[length - 2..];
    let ok = actual.eq_ignore_ascii_case(expected.as_bytes())
        || (actual[0] == b' ' && actual[1].eq_ignore_ascii_case(&expected.as_bytes()[1]));
    if !ok {
        return (
            None,
            Some(format!(
                "Checksum mismatch (expected {expected:?}, got {:?})",
                String::from_utf8_lossy(actual)
            )),
        );
    }
    let message = match std::str::from_utf8(&data[sep_index..length - 3]) {
        Ok(s) => s,
        Err(_) => return (None, Some("Invalid ASCII encoding".into())),
    };
    let fields: Vec<String> = message.split('|').skip(1).map(|s| s.to_string()).collect();
    (Some(fields), None)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn roundtrip_request() {
        let fields = vec![
            "HMG-50".to_string(),
            "AABBCCDDEEFF".to_string(),
            "HME-4".to_string(),
            "112233445566".to_string(),
            "A".to_string(),
            "120".to_string(),
        ];
        let payload = build_payload(&fields);
        assert_eq!(payload[0], SOH);
        assert_eq!(payload[1], STX);
        let (parsed, err) = parse_message(&payload);
        assert!(err.is_none(), "{err:?}");
        assert_eq!(parsed.unwrap(), fields);
    }

    #[test]
    fn rejects_short() {
        let (p, e) = parse_message(&[0x01, 0x02]);
        assert!(p.is_none());
        assert!(e.unwrap().contains("Too short"));
    }
}

File: esphome/components/ct002/protocol.cpp
#include "protocol.h"

#include <cctype>
#include <cstdio>
#include <cstring>

namespace esphome {
namespace ct002 {

const char *const RESPONSE_LABELS[24] = {
    "meter_dev_type",  "meter_mac_code",  "hhm_dev_type",   "hhm_mac_code",
    "A_phase_power",   "B_phase_power",   "C_phase_power",  "total_power",
    "A_chrg_nb",       "B_chrg_nb",       "C_chrg_nb",      "ABC_chrg_nb",
    "wifi_rssi",       "info_idx",        "x_chrg_power",   "A_chrg_power",
    "B_chrg_power",    "C_chrg_power",    "ABC_chrg_power", "x_dchrg_power",
    "A_dchrg_power",   "B_dchrg_power",   "C_dchrg_power",  "ABC_dchrg_power",
};

uint8_t calculate_checksum(const uint8_t *data, size_t len) {
  uint8_t xor_val = 0;
  for (size_t i = 0; i < len; ++i) {
    xor_val ^= data[i];
  }
  return xor_val;
}

size_t compute_length(size_t payload_without_length_size) {
  // Mirrors Python: base = SOH + STX + body + ETX + checksum(2 bytes).
  // Caller passes the size of body only (after SEPARATOR is prepended) —
  // we add the 1+1+1+2 framing bytes here ourselves to match the Python
  // contract where `payload_without_length` is the joined message bytes.
  const size_t base_size = 1 + 1 + payload_without_length_size + 1 + 2;
  for (size_t length_digits = 1; length_digits <= 4; ++length_digits) {
    const size_t total = base_size + length_digits;
    char buf[8];
    const int written = std::snprintf(buf, sizeof(buf), "%zu", total);
    if (written > 0 && static_cast<size_t>(written) == length_digits) {
      return total;
    }
  }
  return 0;
}

std::vector<uint8_t> build_payload(const std::vector<std::string> &fields) {
  std::string message_str;
  for (const auto &f : fields) {
    message_str.push_back(SEPARATOR);
    message_str.append(f);
  }
  const size_t total = compute_length(message_str.size());
  if (total == 0) {
    return {};
  }
  std::vector<uint8_t> payload;
  payload.reserve(total);
  payload.push_back(SOH);
  payload.push_back(STX);
  char len_buf[8];
  const int len_written = std::snprintf(len_buf, sizeof(len_buf), "%zu", total);
  for (int i = 0; i < len_written; ++i) {
    payload.push_back(static_cast<uint8_t>(len_buf[i]));
  }
  for (char c : message_str) {
    payload.push_back(static_cast<uint8_t>(c));
  }
  payload.push_back(ETX);
  const uint8_t checksum = calculate_checksum(payload.data(), payload.size());
  char cs_buf[3];
  std::snprintf(cs_buf, sizeof(cs_buf), "%02x", checksum);
  payload.push_back(static_cast<uint8_t>(cs_buf[0]));
  payload.push_back(static_cast<uint8_t>(cs_buf[1]));
  return payload;
}

static void set_error(std::string *out, const char *msg) {
  if (out != nullptr) {
    *out = msg;
  }
}

std::optional<std::vector<std::string>> parse_request(const uint8_t *data, size_t len,
                                                      std::string *error_out) {
  if (len < 10) {
    set_error(error_out, "Too short");
    return std::nullopt;
  }
  if (data[0] != SOH || data[1] != STX) {
    set_error(error_out, "Missing SOH/STX");
    return std::nullopt;
  }
  size_t sep_index = 0;
  bool found_sep = false;
  for (size_t i = 2; i < len; ++i) {
    if (data[i] == static_cast<uint8_t>(SEPARATOR)) {
      sep_index = i;
      found_sep = true;
      break;
    }
  }
  if (!found_sep) {
    set_error(error_out, "No separator after length");
    return std::nullopt;
  }
  size_t length = 0;
  for (size_t i = 2; i < sep_index; ++i) {
    const uint8_t b = data[i];
    if (b < '0' || b > '9') {
      set_error(error_out, "Invalid length field");
      return std::nullopt;
    }
    length = length * 10 + (b - '0');
  }
  if (length != len) {
    set_error(error_out, "Length mismatch");
    return std::nullopt;
  }
  if (data[len - 3] != ETX) {
    set_error(error_out, "Missing ETX");
    return std::nullopt;
  }
  const uint8_t xor_val = calculate_checksum(data, len - 2);
  char expected[3];
  std::snprintf(expected, sizeof(expected), "%02x", xor_val);
  const uint8_t actual_hi = data[len - 2];
  const uint8_t actual_lo = data[len - 1];
  const bool exact = (static_cast<char>(std::tolower(actual_hi)) == expected[0]) &&
                     (static_cast<char>(std::tolower(actual_lo)) == expected[1]);
  // Real-firmware quirk: tolerate a leading space in the high nibble.
  const bool space_tolerant = (actual_hi == ' ') &&
                              (static_cast<char>(std::tolower(actual_lo)) == expected[1]);
  if (!exact && !space_tolerant) {
    set_error(error_out, "Checksum mismatch");
    return std::nullopt;
  }
  std::vector<std::string> fields;
  std::string current;
  // Python: `data[sep_index:-3].decode("ascii").split("|")[1:]`. data[sep_index]
  // is the leading '|' so skip it; iterate through the byte before ETX.
  for (size_t i = sep_index + 1; i < len - 3; ++i) {
    const uint8_t b = data[i];
    if (b >= 0x80) {
      set_error(error_out, "Invalid ASCII encoding");
      return std::nullopt;
    }
    if (b == static_cast<uint8_t>(SEPARATOR)) {
      fields.push_back(std::move(current));
      current.clear();
    } else {
      current.push_back(static_cast<char>(b));
    }
  }
  fields.push_back(std::move(current));
  return fields;
}

}  // namespace ct002
}  // namespace esphome
